csrc/optimization/exact_mapped_extent_substitution.cpp

```diff
@@ -0,0 +1,86 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <debug.h>
+#include <ir/utils.h>
+#include <optimization/exact_mapped_extent_substitution.h>
+#include <options.h>
+#include <root_domain_map.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+// Skip broadcast without expanded extent
+// Skip derived domains, e.g. iS11{( i0 * i2 )}rf
+// Skip domain whose extent is derived e.g. iS12{( i0 * i2 )}
+// e.g. in this set { iS11{( i0 * i2 )}rf; iS12{( i0 * i2 )}; iS14{i3} } from
+// NVFuserTest.SymbolicSqueeze, we can't substitute {i0 * i2} with {i3},
+// otherwise, ValidateDomainEquivalence fails. If we really want to substitute,
+// we may need to skip or modify ValidateDomainEquivalence.
+inline bool isNonSubstitutableID(const IterDomain* id) {
+  return (id->isBroadcast() && !id->hasExpandedExtent()) || id->definition() ||
+      id->getMaybeExpandedExtent()->definition();
+}
+
+void exactMappedExtentSubstitution(Fusion* fusion) {
+  // map non-const extents to const extents
+  std::unordered_map<Val*, Val*> replacement_map;
+
+  const auto mapped_sets = ExactRootDomainMap(fusion).getMappedSets();
+  // Loop over each exact root domain set
+  for (const auto& set_ptr : mapped_sets.disjointSets()) {
+    // (1) pick a const extent
+    // (2) if no const extent, pick the var with the lowest name()
+    Val* const_extent = nullptr;
+    Val* lowest_val = nullptr;
+    for (auto id : *set_ptr) {
+      if (isNonSubstitutableID(id)) {
+        continue;
+      }
+      // find the const extent
+      if (id->getMaybeExpandedExtent()->isConstScalar()) {
+        const_extent = id->getMaybeExpandedExtent();
```

Comment:
Should we check here whether `const_extent` is already set and, if so, check that the values match? My concern is that we might use two conflicting values that are exact mapped with one another. I think if this happens the error comes from `ExpressionEvaluator`, but here we are remapping those values so later evaluation won't show any conflicts. This type of error would likely be caught at concretization, but it would be pretty cheap to just hold onto the const value if it exists here in order to do a check. What do you think?

Reply:
This additional verification can be smoothly integrated into the existing code. Excellent idea!